File: src/main/java/com/airplanegobrr/blockbreakerblockplacer/events/onBlockDispense.java
package com.airplanegobrr.blockbreakerblockplacer.events;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.Dispenser;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockDispenseEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.block.data.Directional;

import com.airplanegobrr.blockbreakerblockplacer.Blockbreakerblockplacer;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.function.Supplier;

public class onBlockDispense implements Listener {
    private final Blockbreakerblockplacer main;

    public onBlockDispense(Blockbreakerblockplacer main) {
        this.main = main;
    }

    @EventHandler
    public void BlockDispenseEvent(BlockDispenseEvent event) {
        Block block = event.getBlock();
        Location loc = block.getLocation();
        String xyzName = (loc.getBlockX() + "," + loc.getBlockY() + "," + loc.getBlockZ());
        ItemStack item = event.getItem();
        Material itemType = item.getType();

        // Config stuff
        main.getConfig().addDefault("debug", false);
        boolean debug = main.getConfig().getBoolean("debug");

        main.getConfig().addDefault("blockBreakers." + xyzName, false);
        boolean bbEnabled = main.getConfig().getBoolean("blockBreakers." + xyzName);

        main.getConfig().addDefault("blockPlacers." + xyzName, false);
        boolean bpEnabled = main.getConfig().getBoolean("blockPlacers." + xyzName);

        // Real stuff

        // blockbreaker
        if (bbEnabled) {
            if (debug) {
                main.getLogger().info(item.toString());
                main.getLogger().info(item.getType().toString());
                main.getLogger().info(item.getType().name());
                main.getLogger().info(xyzName);
            }

            if (item.getType().toString().contains("PICKAXE")) {
                if (debug) {
                    main.getLogger().info("Pickaxe detected!");
                }
                BlockFace bsFace = ((Directional) block.getBlockData()).getFacing();
                Block bsBlock = block.getRelative(bsFace);
                Material bsMaterial = bsBlock.getType();
                // if the material is not bedrock then break it
                if (bsMaterial != Material.BEDROCK) {
                    if(bsMaterial == Material.AIR || bsMaterial == Material.WATER || bsMaterial == Material.LAVA) {
                        event.setCancelled(true);
                        return;
                    }
                    bsBlock.breakNaturally();
                    try {
                        event.setCancelled(true);
                        return;
                    } catch (Exception e) {
                        if (debug) {
                            main.getLogger().info("Failed to cancel event!");
                            main.getLogger().warning((Supplier<String>) e);
                        }
                        return;
                    }
                }
            }
        }

        // blockplacer
        if (bpEnabled) {
            if (debug) {
                main.getLogger().info(item.toString());
                main.getLogger().info(item.getType().toString());
                main.getLogger().info(item.getType().name());
                main.getLogger().info(xyzName);
            }

            if (itemType.isBlock()){
                BlockFace bsFace = ((Directional) block.getBlockData()).getFacing();
                Block bsBlock = block.getRelative(bsFace);
                Material bsMaterial = bsBlock.getType();

                if (bsMaterial == Material.AIR || bsMaterial == Material.WATER || bsMaterial == Material.LAVA) {
                    bsBlock.setType(itemType);
                    event.setCancelled(true);

                    Block blockUpdated = event.getBlock();
                    Dispenser dispenserUpdated = (Dispenser) blockUpdated.getState();
                    ItemStack itemUpdated = event.getItem();

                    //remove item from dispenser with a BukkitRunnable
                    new BukkitRunnable() {
                        @Override
                        public void run() {
                            if (debug) {
                                main.getLogger().info("Removing item from dispenser");
                            }
                            dispenserUpdated.getInventory().removeItem(itemUpdated);
                        }
                    }.runTaskLater(main, 1);
                }
            }
        }

    }
}

File: src/main/kotlin/xyz/airplanegobrr/blockbreakerplacer/onBlockDispense.kt
package xyz.airplanegobrr.blockbreakerplacer

import org.bukkit.Material
import org.bukkit.block.Dispenser
import org.bukkit.block.data.Directional
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockDispenseEvent
import org.bukkit.scheduler.BukkitRunnable
import java.util.function.Supplier

class onBlockDispense(main: Main) : Listener {
    private val main: Main

    init {
        this.main = main
        main.logger.info("Event is listening")
    }
    @EventHandler
    fun BlockDispenseEvent(event: BlockDispenseEvent) {
        val block = event.block
        val loc = block.location
        val xyzName: String = "${loc.world?.name}${loc.blockX},${loc.blockY},${loc.blockZ}"
        val item = event.item
        val itemType = item.type

        // main.config.load()
        val debug: Boolean = main.config.getBoolean("debug")
        val bbEnabled: Boolean = main.config.getBoolean("breakers.$xyzName")
        val bpEnabled: Boolean = main.config.getBoolean("placers.$xyzName")

        if (debug) main.logger.info("Checking $xyzName $bbEnabled $bpEnabled $debug")
        // Real stuff

        var noBreak = arrayOf(Material.WATER, Material.LAVA, Material.AIR)

        // blockbreaker
        if (bbEnabled) {
            if (debug) {
                main.logger.info(item.toString())
                main.logger.info(item.type.toString())
                main.logger.info(item.type.name)
                main.logger.info(xyzName)
            }
            if (item.type.toString().contains("PICKAXE")) {
                if (debug) main.logger.info("Pickaxe detected!")

                val bsFace = (block.blockData as Directional).facing
                val bsBlock = block.getRelative(bsFace)
                val bsMaterial = bsBlock.type
                // if the material is not bedrock then break it
                if (bsMaterial != Material.BEDROCK && !noBreak.contains(bsMaterial)) {
                    try {
                        bsBlock.breakNaturally()
                        event.isCancelled = true
                        return
                    } catch (e: Exception) {
                        if (debug) {
                            main.logger.info("Failed break block!")
                            main.logger.warning(e as Supplier<String?>)
                        }
                        return
                    }
                } else {
                    event.isCancelled = true
                }
            }
        }

        var noPlace = arrayOf(Material.WATER, Material.LAVA, Material.AIR)

        // blockplacer
        if (bpEnabled) {
            if (debug) {
                main.logger.info(item.toString())
                main.logger.info(item.type.toString())
                main.logger.info(item.type.name)
                main.logger.info(xyzName)
            }
            if (itemType.isBlock) {
                val bsFace = (block.blockData as Directional).facing
                val bsBlock = block.getRelative(bsFace)
                val bsMaterial = bsBlock.type
                main.logger.info(bsBlock.type.name)
                if (noPlace.contains(bsMaterial)) {
                    main.logger.info("Passed check!")
                    bsBlock.type = itemType
                    event.isCancelled = true
                    val blockUpdated = event.block
                    val dispenserUpdated = blockUpdated.state as Dispenser
                    val itemUpdated = event.item

                    //remove item from dispenser with a BukkitRunnable
                    object : BukkitRunnable() {
                        override fun run() {
                            if (debug) main.logger.info("Removing item from dispenser")

                            dispenserUpdated.inventory.removeItem(itemUpdated)
                        }
                    }.runTaskLater(main, 1)
                } else {
                    event.isCancelled = true
                }
            }
        }
    }
}